tests/unit/argument-test.js

```diff
@@ -1,193 +1,409 @@
+          });
+        }
 
-    assert.equal(fooWithValues.bar, 3, 'argument default can be overriden');
-    assert.equal(fooWithValues.baz, 2, 'class field default cannot be overriden');
-  });
+        @argument('number')
+        prop = 123;
+      }
 
-  test('it works with defaultIfUndefined and native getters', function(assert) {
-    class Foo extends EmberObject {
-      @argument({ defaultIfUndefined: true })
-      bar = 1;
-    }
+      const foo = Foo.create();
+
+      // Works by default
+      assert.equal(foo.get('prop'), 123, 'default value provided');
+
+      // Works when dependent key is set
+      foo.set('prop', 456);
+      assert.equal(foo.get('prop'), 456, 'can set dependent key');
+    });
+
+    test('typed native setter does not trigger property notifications if value is unchanged', function(assert) {
+      let value = 0;
+
+      class Foo extends EmberObject {
+        @argument('number')
+        get prop() {
+          return 123;
+        }
 
-    const foo = Foo.create({ bar: undefined });
-    const fooWithValues = Foo.create({ bar: 3 });
+        set prop(value) {
+          // do nothing
+        }
 
-    assert.equal(foo.bar, 1, 'argument default gets set correctly');
-    assert.equal(fooWithValues.bar, 3, 'argument default can be overriden');
+        @computed('prop')
+        get otherProp() {
+          return value++;
+        }
+      }
 
-    foo.set('bar', undefined);
-    assert.equal(foo.bar, 1, 'argument cannot be set to undefined in repeated usage');
-  });
+      const foo = Foo.create();
 
-  test('it works with defaultIfNullish and native getters', function(assert) {
-    class Foo extends EmberObject {
-      @argument({ defaultIfNullish: true })
-      bar = 1;
-    }
+      assert.equal(foo.get('otherProp'), 0, 'computed is correct before');
 
-    const fooWithUndefined = Foo.create({ bar: undefined });
-    const fooWithNull = Foo.create({ bar: null });
-    const fooWithValues = Foo.create({ bar: 3 });
+      foo.set('prop', 123);
 
-    assert.equal(fooWithUndefined.bar, 1, 'argument default gets set correctly');
-    assert.equal(fooWithNull.bar, 1, 'argument default gets set correctly');
-    assert.equal(fooWithValues.bar, 3, 'argument default can be overriden');
+      assert.equal(foo.get('otherProp'), 0, 'computed did not change');
+    });
+  });
 
-    fooWithUndefined.set('bar', null);
-    fooWithNull.set('bar', undefined);
-    assert.equal(fooWithUndefined.bar, 1, 'argument cannot be set to null in repeated usage');
-    assert.equal(fooWithNull.bar, 1, 'argument cannot be set to undefined in repeated usage');
+  module('compatibility with native classes', function() {
+    // Skipped because the native getter/setter in the parent class is currently
+    // clobbered by the definition of the instance property in the constructor
+    // of the base class
+    skip('typed value can be provided by getter/setter', function(assert) {
```

Comment:
I added the context above, but want to call it out explicitly here, too:

I wrestled with this for a few hours to try to understand what's going on and why it didn't work. I think it comes down to this sequence of events:

- `@argument('number') prop` has the behavior of running `Object.defineProperty` at the end of the constructor
- The `Object.defineProperty` does not invoke the parent class `getter` or `setter`; the property is overwritten, not set. Because of this, the native `getter` and `setter` are entirely ignored
- `prop` is interpreted, during initial validation, to be `undefined` and thus fails the check

This has nothing to do with the code in this library; I validated all of this behavior in a fresh Babel project to understand what is happening.

I think the best course of action is just to not support this use-case at all, because anyone that has a `getter` and a property initializer is going to run into this same problem, `@argument` or not. I'm not sure what we could do to attempt to avoid the problem, but would love to re-enable this test at some point if we can figure out the right way to go about it. I left the test as a `skip` as an explicit reminder of this behavior.